### PR TITLE
Refactor/Intro

### PR DESCRIPTION
**Pull request by maktarsis**

### CODE REVIEW

File: libs/intro/generate/generate-thumbnails.ts [hunk 1]
```diff
@@ -8,100 +8,138 @@ import {
...
+const slides$ = gSlides$.pipe(
+  switchMap((gSlides: GSlides) =>
+    from(
+      gSlides.presentations.get({
+        presentationId: CODELAB_PRESENTATION_ID
+      })
+    ).pipe(map(response => response.data.slides.map(slide => slide.objectId)))
+  ),
+  withLatestFrom(gSlides$),
+  switchMap(([objectIds, gSlides]: [string[], GSlides]) =>
```
Comment (NothingEverHappens) [inline libs/intro/generate/generate-thumbnails.ts:57]: just move it to the previous switchMap and drop withLatestFrom

File: libs/intro/generate/generate-thumbnails.ts [hunk 2]
```diff
@@ -8,100 +8,138 @@ import {
...
+      output: process.stdout
     })
-  );
+  ),
+  switchMap(
+    (readline): Observable<string> => {
+      return new Observable(observer =>
+        readline.question('Enter the code from that page here: ', code => {
+          readline.close();
+          observer.next(code);
```
Comment (NothingEverHappens) [inline libs/intro/generate/generate-thumbnails.ts:113]: complete

File: libs/intro/generate/generate-thumbnails.ts [hunk 3]
```diff
@@ -8,100 +8,138 @@ import {
...
+        if (err) {
+          observer.error(err);
+          return console.error('Error retrieving access token', err);
+        }
 
-  if (!existsSync(ASSETS_SLIDES_PATH)) {
-    mkdirSync(ASSETS_SLIDES_PATH);
-  }
+        oAuth2Client.setCredentials(token);
+        observer.next(token);
```
Comment (NothingEverHappens) [inline libs/intro/generate/generate-thumbnails.ts:128]: complete

File: libs/intro/generate/generate-thumbnails.ts [hunk 4]
```diff
@@ -8,100 +8,138 @@ import {
...
+  finalize((): void => console.log('Completed ✅ '))
+);
 
-  await outputJSON(
-    SLIDES_METADATA_PATH,
-    { count: slides.length, contentUrls },
-    { spaces: 2 }
-  );
-}
+token$.pipe(take(1)).subscribe(() => {});
```
Comment (NothingEverHappens) [inline libs/intro/generate/generate-thumbnails.ts:145]: While I think overall using rxjs improves things here, I find the structure a bit confusing:

Lots of work happens as a side effects (in tap)
And the actual work happens by subscribing to a token$ observable? Why?

File: libs/intro/generate/helpers/slides.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  readJSON(TOKEN_PATH)
+).pipe(
+  take(1),
+  withLatestFrom(oAuth2Client$),
+  tap(([token, oAuth2Client]) => {
+    oAuth2Client.setCredentials(token);
+  }),
+  catchError(() => getNewToken$),
+  withLatestFrom(readline$),
+  tap(([, readline]) => readline.close()),
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/slides.ts:53]: This is weird, why do you need to close it here? Looks like an implementation detail

File: libs/intro/generate/helpers/slides.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  switchMap(() => fetchSlides$)
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/slides.ts:54]: switchMapTo?

File: libs/intro/generate/helpers/utils.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    const authUrl = oAuth2Client.generateAuthUrl({
+      access_type: 'offline',
+      scope: SCOPE
+    });
+    console.log('Authorize this app by visiting this url:', authUrl);
+  }),
+  withLatestFrom(readline$),
+  switchMap(([, readline]) =>
+    bindCallback(
+      readline.question.bind(readline, 'Enter the code from that page here: ')
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/utils.ts:52]: Should you readline.close here?

File: libs/intro/generate/generate-thumbnails.ts [hunk 8]
```diff
@@ -1,107 +1,33 @@
...
+
+getSlides$.subscribe(
+  async (contentUrls: string[]): Promise<void> => {
+    if (!existsSync(ASSETS_SLIDES_PATH)) {
+      mkdirSync(ASSETS_SLIDES_PATH);
+    }
+
+    let i = 0;
+
+    for await (const url of contentUrls) {
```
Comment (NothingEverHappens) [inline libs/intro/generate/generate-thumbnails.ts:16]: nit/optional: we can use promise all and parallelize it

File: libs/intro/generate/helpers/slides.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+              })
+            ).pipe(map(response => response.data.contentUrl))
+        );
+        return forkJoin(thumbnailRequests);
+      })
+    )
+  )
+);
+
+export const getSlides$ = from<ObservableInput<Token>>(
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/slides.ts:43]: You don't need this whole getSlides$ thing, all this magic should happen in oAuth2Client$ observable

File: libs/intro/generate/helpers/slides.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  tap,
+  withLatestFrom
+} from 'rxjs/operators';
+
+import { GSlides, Token } from './types';
+import { CODELAB_PRESENTATION_ID, TOKEN_PATH } from './const';
+import { oAuth2Client$, getNewToken$ } from './utils';
+
+export const fetchSlides$ = oAuth2Client$.pipe(
+  map((auth: OAuth2Client): GSlides => google.slides({ version: 'v1', auth })),
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/slides.ts:19]: This file doesn't need to know anything about the auth, it only needs gSlides$ and pipe through it

File: libs/intro/generate/helpers/utils.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  defer,
+  from,
+  ObservableInput,
+  of
+} from 'rxjs';
+
+import { Credentials } from './types';
+import { CREDENTIALS_PATH, SCOPE, TOKEN_PATH } from './const';
+
+export const oAuth2Client$ = from<ObservableInput<Credentials>>(
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/utils.ts:25]: optional: In total honesty seems like readJSONSync would be way cleaner here

File: libs/intro/generate/helpers/slides.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import { defer, forkJoin, from, ObservableInput } from 'rxjs';
+import { map, switchMap, switchMapTo } from 'rxjs/operators';
+
+import { CODELAB_PRESENTATION_ID } from './const';
+import { gSlides, validateToken$ } from './utils';
+
+export const getSlides$ = validateToken$.pipe(
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/slides.ts:7]: This is a bit clunky - use validateToken$, and then gSlides, why not export just single observable - gSlides$?

File: libs/intro/generate/helpers/slides.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  switchMapTo(
+    defer(() =>
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/slides.ts:9]: switchMap + defer = switchMap, ahaha

File: libs/intro/generate/helpers/slides.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      from(
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/slides.ts:10]: You don't need from, you can return promises in switch map

File: libs/intro/generate/helpers/utils.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+      readline.question.bind(readline, 'Enter the code from that page here: ')
+    )()
+  ),
+  switchMap(code =>
+    bindNodeCallback(oAuth2Client.getToken.bind(oAuth2Client))(code)
+  ),
+  tap(([token]): void => {
+    oAuth2Client.setCredentials(token);
+  }),
+  switchMap(([token]) => from(outputJSON(TOKEN_PATH, token, { spaces: 2 }))),
```
Comment (NothingEverHappens) [inline libs/intro/generate/helpers/utils.ts:65]: don't need from

File: libs/intro/generate/helpers/slides.ts (hunk 13 above)
Comment (maktarsis) [reply libs/intro/generate/helpers/slides.ts:9]: haha
omg really!)

File: libs/intro/generate/generate-thumbnails.ts [hunk 16]
```diff
@@ -1,107 +1,40 @@
...
-    });
-  });
+import {
+  ASSETS_SLIDES_PATH,
+  CREDENTIALS_PATH,
+  SLIDES_METADATA_PATH
+} from './helpers/const';
+import { getSlides$ } from './helpers/slides';
+
+if (!existsSync(CREDENTIALS_PATH)) {
```
Comment (NothingEverHappens) [inline libs/intro/generate/generate-thumbnails.ts:12]: Move this to the file which actually does something with credentials

File: libs/intro/generate/generate-thumbnails.ts [hunk 17]
```diff
@@ -1,107 +1,40 @@
...
-        pageObjectId
-      });
-
-      return response.data.contentUrl;
-    })
-  );
-
-  if (!existsSync(ASSETS_SLIDES_PATH)) {
-    mkdirSync(ASSETS_SLIDES_PATH);
+getSlides$.subscribe(
```
Comment (NothingEverHappens) [inline libs/intro/generate/generate-thumbnails.ts:16]: Can you verify what happens when you don't subscribe? Ideally - nothing - no file read, no requests

Comment (maktarsis) [reply libs/intro/generate/generate-thumbnails.ts:16]: yeah, I'll add additional script for verification

**Pull request merged**